aesara/tensor/elemwise.py

```diff
@@ -1364,41 +1334,41 @@ def make_node(self, input):
         from aesara.tensor.basic import as_tensor_variable
 
         input = as_tensor_variable(input)
+        inp_dims = input.type.ndim
+        inp_bdcast = input.type.broadcastable
+        inp_dtype = input.type.dtype
+        copy_op = False
 
-        if self.axis is not None:
-            for axis in self.axis:
-                if axis >= input.type.ndim or (
-                    axis < 0 and abs(axis) > input.type.ndim
-                ):
-                    raise ValueError(
-                        f"Not enough dimensions on {input} to reduce on axis {axis}"
-                    )
-        input = as_tensor_variable(input)
         axis = self.axis
         if axis is None:
-            axis = list(range(len(input.type.broadcastable)))
-        if any(a < 0 for a in axis):
-            axis2 = []
-            for a in self.axis:
-                if a < 0:
-                    axis2.append(a + input.type.ndim)
-                else:
-                    axis2.append(a)
-            assert len(axis) == len(axis2)
-            axis = tuple(axis2)
-            # We can't call self.__class__() as there is a class that
-            # inherits from CAReduce that doesn't have the same signature
+            axis = list(range(len(inp_bdcast)))
+
+        axis = list(axis)
+        for i, a in enumerate(axis):
+            if a >= inp_dims or a < -inp_dims:
+                raise ValueError(
+                    f"Not enough dimensions on {input} to reduce on axis {a}"
+                )
+            if a < 0:
+                copy_op = True
+                axis[i] = a + inp_dims
+
+        # We can't call self.__class__() as there is a class that
+        # inherits from CAReduce that doesn't have the same signature
+        if copy_op:
             op = copy(self)
             op.set_ufunc(op.scalar_op)
-            op.axis = axis
+            assert len(axis) == len(self.axis)
+            op.axis = tuple(axis)
         else:
             op = self
```

Comment:
We should address this at some point, because creating a temporary `Op` seems so wildly unnecessary.  It also complicates things down the road, because `op(input).owner.op is not op`, which is very unexpected&mdash;especially since other `Op`s don't behave this way.

The main idea seems to be that `op.axis` isn't sufficient when `a < 0` above, so, in order to affect a change in `op.axis`, an entirely new `Op` is created.  Instead, we could easily make `axis` another input (i.e. the return value below becomes `Apply(op, [axis, input], [output])` and the methods that use the inputs list are updated to get `self.axis` from the list).

Reply:
I created an issue for this: #299.